src/jit/emitxarch.cpp

```diff
@@ -2901,174 +2901,282 @@ void emitter::emitInsStoreLcl(instruction ins, emitAttr attr, GenTreeLclVarCommo
 {
-    // dst can only be a reg or modrm
-    assert(!dst->isContained() || dst->isUsedFromMemory() || instrIs3opImul(ins)); // dst on these isn't really the dst
+    // We can only have one memory operand and only src can be a constant operand
+    GenTree* memOp = nullptr;
+    GenTree* cnsOp = nullptr;
+    GenTree* otrOp = nullptr;
 
-#ifdef DEBUG
-    // src can be anything but both src and dst cannot be addr modes
-    // or at least cannot be contained addr modes
-    if (dst->isUsedFromMemory())
+    if (dst->isContained() || (dst->isLclField() && (dst->gtRegNum == REG_NA)) || dst->isUsedFromSpillTemp())
     {
+        // dst can only be a reg or modrm
+        assert(dst->isUsedFromMemory() || (dst->gtRegNum == REG_NA) || instrIs3opImul(ins)); // dst on 3opImul isn't really the dst
         assert(!src->isUsedFromMemory());
-    }
-
-    if (src->isUsedFromMemory())
-    {
-        assert(!dst->isUsedFromMemory());
-    }
-#endif
 
-    // find which operand is a memory op (if any)
-    // and what its base is
-    GenTreeIndir* mem     = nullptr;
-    GenTree*      memBase = nullptr;
+        memOp = dst;
 
-    if (dst->isContainedIndir())
-    {
-        mem = dst->AsIndir();
+        if (src->isContained())
+        {
+            assert(src->IsCnsIntOrI());
+            cnsOp = src;
+        }
+        else
+        {
+            otrOp = src;
+        }
     }
-    else if (src->isContainedIndir())
+    else if (src->isContained() || src->isUsedFromSpillTemp())
     {
-        mem = src->AsIndir();
-    }
+        assert(!dst->isUsedFromMemory());
+        otrOp = dst;
 
-    if (mem)
-    {
-        memBase = mem->gtOp1;
+        if (src->IsCnsIntOrI() || src->IsCnsFltOrDbl())
+        {
+            assert(!src->isUsedFromMemory() || src->IsCnsFltOrDbl());
+            cnsOp = src;
+        }
+        else
+        {
+            assert(src->isUsedFromMemory());
+            memOp = src;
+        }
     }
 
-    // Find immed (if any) - it cannot be the dst
-    // SSE2 instructions allow only the second operand to be a memory operand.
-    GenTreeIntConCommon* intConst = nullptr;
-    GenTreeDblCon*       dblConst = nullptr;
-    if (src->isContainedIntOrIImmed())
-    {
-        intConst = src->AsIntConCommon();
-    }
-    else if (src->isContainedFltOrDblImmed())
+    if (memOp != nullptr)
     {
-        dblConst = src->AsDblCon();
-    }
+        TempDsc* tmpDsc = nullptr;
+        unsigned varNum = BAD_VAR_NUM;
+        unsigned offset = (unsigned)-1;
 
-    // find local field if any
-    GenTreeLclFld* lclField = nullptr;
-    if (src->isLclFldUsedFromMemory())
-    {
-        lclField = src->AsLclFld();
-    }
-    else if (dst->isLclField() && dst->gtRegNum == REG_NA)
-    {
-        lclField = dst->AsLclFld();
-    }
+        if (memOp->isIndir())
+        {
+            GenTreeIndir* memIndir = memOp->AsIndir();
+            GenTree*      memBase  = memIndir->gtOp1;
```

Comment:
I kept the old name here, but I don't think `memBase` is appropriate.

There is also `mem->Base()` which returns something slightly different and which is used in `emitHandleMemOp`